### ISSUE TITLE
pom.xml for a jar/uberjar in a monorepo has no dependencies

### ISSUE THREAD

**Issue by golanw** [seancorfield/depstar#78]
I'm using a monorepo structure as described in this blog post: https://corfield.org/blog/2021/02/23/deps-edn-monorepo/

Since all commands are being executed from the top-level folder and libs are located in nested folders, this creates a problem when generating pom.xml files for a jar/uberjar:

Let's assume commands are being executed from: `/home/myproject/clojure`
And let's assume library name is "mylib" and is located here: `/home/myproject/clojure/mylib`

When trying to run this from the top-level folder:
`clojure -X:jar ... :aliases '[:defaults :mylib]' ...`
- The jar/uberjar is being created with the correct content
- However pom.xml doesn't contain any dependencies

Note: A similar empty pom.xml is being created when trying to run from the top-level folder:
`clojure -A:defaults:mylib -Spom`
and a warning is printed:
```
Skipping coordinate: {:local/root /home/myproject/clojure/mylib, .....
```

The problem is related to `read-edn-files` function or maybe to tools.deps `find-edn-maps`.
I understand It makes sense to ignore `local/root` dependencies when generating pom.xml
but I think a good solution will be a flag to allow the user to decide whether he wants to ignore these `local/root` deps, or actually traverse into them and add their dependencies to the final pom.xml

**Comment by seancorfield** [comment]
This is definitely a tricky one and not a use case `depstar` is designed to solve directly.

First off, you don't need a `pom.xml` file for an uberjar, just to clarify that.

Second, you _do_ need a `pom.xml` file for a library JAR so that's the scenario I'll address here.

If you're using the monorepo structure I outlined in my blog post, then you have a repo-level `deps.edn` at the top of the tree and each subproject has its own `deps.edn` files. The dependencies of each subproject should be listed in its own `deps.edn` file -- but from the repo-level's point of view, those dependencies are _transitive_ dependencies because the subprojects are being treated like libraries (just in source form). The `pom.xml` for a subproject-as-library would need to be created _against that subproject_ not against the repo-level `deps.edn` because you want the context of the subproject's dependencies directly.

Essentially, you need to run `depstar` _inside_ the subproject so that it sees the subproject's dependencies as primary, rather than transitive.

If you're relying heavily on `:override-deps` via a `:defaults` alias in the repo-level `deps.edn` file, that's going to be problematic -- because there's no context where you can both get the subproject's direct dependencies _and_ get the `:defaults` alias in play from the repo-level `deps.edn` file.

Polylith addresses this in an interesting way by having explicit `projects` for things you want to build and having a project-specific `deps.edn` file focused on building an artifact, so you can specify the dependencies directly as needed -- although that can mean duplication of those dependencies between the subproject (`base`, generally) and the `project` itself. For building libraries out of a monorepo, that seems like a reasonable compromise. Polylith has no support for "pinning" versions right now, but it's something I've been discussing with Joakim and he's thinking about how best to solve that in the context of Polylith's existing structure (at least, the `issue-66` branch structure, which is what I've been tracking -- both in `clj-new` and at work).

For what it's worth, we recently decided to abandon `:override-deps` and our `:defaults` alias specifically because of problems such as these, so we've made the trade-off of accepting some duplication of dependency versions across multiple `deps.edn` files in return for a structure that simplifies our tool usage: we can now do most operations directly inside our subprojects pretty much as easily as we could run them from the repo-level directory. A blog post is coming soon about that.

Regarding `:local/root` (and `:git/url`) dependencies, `depstar` already has an option to use `:paths-only true` to avoid dragging in `:deps`/`:extra-deps` specified in source form. There are good reasons on both sides of that choice: if you're building an artifact that depends on sources -- either local projects or git-based projects -- you may well _need_ those sources included in your JAR because they aren't published to an accessible Maven-style repo (which is why including them is the default, but you can opt-out of that).

So the TL;DR is that "solving" this in the context of a repo-level `deps.edn` file (like ours at work or like Polylith) is out of scope for `depstar` -- and my advice is to build the `pom.xml` file and JAR for a library subproject directly _inside_ the subproject and do what you need with the subproject's `deps.edn` file to make that work, at the cost of duplicating some dependency version specifications.

**Comment by golanw** [comment]
Thank you for the detailed answer!

I agree about the uberjars - it's only a "cosmetic" issue there.

Regarding JARs, Indeed this is how I solve it.
I already had to duplicate some deps because Cursive scans each project on its own and doesn't merge the empty maps well.
However instead of running `-X:jar` from each project level, I still run it from top-level, using a new flag in I added in depstar (in a fork), that allows me to specify the project's `deps.edn` path to be taken instead of the default one (which is in current working dir).
Basically I'm passing this path to `find-edn-maps` instead of using the no-arg version of the function:
See https://github.com/clojure/tools.deps.alpha/blob/master/src/main/clojure/clojure/tools/deps/alpha.clj#L116
